### PR TITLE
Caldav/refactor: Mutiple Fixes and General code improvement

### LINKED ISSUE #5570: Meetings in Apple calendar are offset when calculating available time slots
Found a bug? Please fill out the sections below. 👍

The calendar I use is iCloud and I am located in continental Europe (CET). My Calendar is in this timezone on my device. I have confirmed on [icloud.com](http://icloud.com/) that it is the same there. When an event is scheduled via [cal.com](http://cal.com/), it is correctly scheduled at the right time. However, the events in my calendar that are checked for conflicts are not taken into account correctly. They are offset by 1 hour. It's as if an event set in the calendar outside of [cal.com](http://cal.com/) for 10:00 CET is in fact considered to be at 11:00 CET, thus defeating the purpose of the app, since double bookings are possible. Events which are schedules within [cal.com](http://cal.com/), however, are correctly taken into account.

A summary of the issue. This needs to be a clear detailed-rich summary.

### Steps to Reproduce

1. I opened a new account
2. I added two Apple calendars, with preexisting meetings, and both set to Central European Time (CET)
3. I created an event type limited to 3 days in the future (for an event) and with an availability schedule for 10:00-19:00 every day
4. Then I went to the booking page
5. The unavailable time slots (due to a preexisting meeting) are offset by one hour on the booking page.
6. I can book a meeting during the preexisting meeting, resulting in a collision
7. That meeting appears correctly in the Apple calendar
8. That meeting also results in correct unavailable time slots

Thus, almost everything seems fine, apart from the fact that the preexisting events are incorrectly scheduled when calculating the available or unavailable time slots. The offset is 1 hour, which is the current time difference between my time zone and UTC - that could be the issue, perhaps.

### Technical details

OS : Mac OS and iOS (tried on 2 devices)
Browsers : Chrome 107 and Firefox 107 (tried both)
Calendar timezone : CET

### LINKED ISSUE #7796: Conflict checks stopped working
### Issue Summary

I am using CalDAV with mailbox.org and cal.com is no longer recognizing recurring events from mailbox.org as blocked. So now I need to reschedule a lot of things :/

This worked before!

### Steps to Reproduce

1. Go to https://www.heylogin.com/call/dominik
2. You'll see that you can book Fridays in my calendar, but Fridays are fully blocked with recurring events
3. Individual events (for example 17-19th April) are recognized correctly

image

### PR DESCRIPTION
**Pull request by alishaz-polymath**
### Refactoring the CalDAV and Apple Calendar code to improve overall stability and support for various caldav services

- Handle failure better so that it doesn't stop the conflict checks for other working integrations
- Improves the Sanitization function to  match the colons, semicolons, and equal signs only when they are followed by whitespace and then immediately followed by another colon, semicolon, or equal sign on the next line. This is an improvement on the current sanitizer function making it broader spectrum and avoiding errors that caused Shared Apple Calendar (other's ownership) and a few other CalDAV service providers to fail. In general, this should allow support for more CalDAV service providers.
- We handle the expanded recurring events now, which should block off all the occurrences of the expanded recurring event in the given timeframe
- Some CalDAV service providers fails silently when we use `EXPAND` property in the fetching of the Calendar Objects (eg. Zoho) and they do this in the back-end. Since there is no way to know for sure if `EXPAND` property is supported or not, we first send the request **with** the `EXPAND` property and then if that returns empty (silent failure) we make a call without the `EXPAND` property.
- We assume that any ICAL object returned without any extension is a `.ics` file and treats it as such. This adds support to the Service Providers which send valid ICAL objects without an extension.

Fixes #7796 #7570 #5604 #5238 #8098 #8166 

### To follow up:

- Improve the timezone registration when timezone is missing
- DST compatibility
- Tests when adding a CalDAV integration

Follow ups issues to fix: #5570 

**Environment**: Staging(main branch)

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- Chore (refactoring code, technical debt, workflow improvements)

## How should this be tested?

- [x] Connect with a shared Apple Calendar with someone else's ownership. It should now work.
- [x] Connect a failed integration. It shouldn't stop working calendar integrations from conflict checking anymore.
- [x] Tested Support for various CalDAV integrations from Vates (Bluemind) to Qloud MSP (NextCloud) and Zoho. Although, reschedule in Zoho isn't working but that's on their end, and the issue is being checked by their team. 

## Checklist

- I haven't performed a self-review of my own code and corrected any misspellings
- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: packages/lib/CalendarService.ts [hunk 1]
```diff
@@ -300,133 +303,127 @@ export default abstract class BaseCalendarService implements Calendar {
...
-        standard.addPropertyWithValue("tzoffsetfrom", tzoffsetfrom);
-        standard.addPropertyWithValue("tzoffsetto", tzoffsetto);
-        // provide a standard dtstart
-        standard.addPropertyWithValue("dtstart", "1601-01-01T00:00:00");
-        timezoneComp.addSubcomponent(standard);
-        vcalendar.addSubcomponent(timezoneComp);
+      let vcalendar: ICAL.Component;
+      try {
+        const jcalData = ICAL.parse(sanitizeCalendarObject(object));
+        vcalendar = new ICAL.Component(jcalData);
```
Comment (alishaz-polymath) [inline packages/lib/CalendarService.ts:319]: We place ICAL.parse in a `try-catch` as .parse function throws error when the object is not of iCalendar 2.0 standard, and it ends up breaking the working calendars as well. The `try-catch` fails it without breaking the flow.

File: packages/lib/CalendarService.ts [hunk 2]
```diff
@@ -300,133 +303,127 @@ export default abstract class BaseCalendarService implements Calendar {
...
-      const vtimezone = vcalendar.getFirstSubcomponent("vtimezone");
-
-      if (event.isRecurring()) {
-        let maxIterations = 365;
-        if (["HOURLY", "SECONDLY", "MINUTELY"].includes(event.getRecurrenceTypes())) {
-          console.error(`Won't handle [${event.getRecurrenceTypes()}] recurrence`);
-          return;
+      // const vevent = vcalendar.getFirstSubcomponent("vevent");
+      const vevents = vcalendar.getAllSubcomponents("vevent");
+      vevents.forEach((vevent) => {
```
Comment (alishaz-polymath) [inline packages/lib/CalendarService.ts:326]: A single Calendar Object might contain multiple events when we're talking about expanded recurring events in particular. Placing them in a forEach ensures we don't miss the next events. This fixes the earlier issue where only the first occurrence of a recurring event was being considered for the freebusy check.

**Review by alishaz-polymath — COMMENTED** [review]
Self Review added

**Review by leog — APPROVED** [review]
Code looks good, approving.

**Pull request merged**